mini-cloudinary.rb

```diff
@@ -16,56 +16,57 @@ def parse_error_to_json(code, message)
-        request = Net::HTTP.new(uri.host, uri.port)
-        request.use_ssl = (uri.scheme == 'https')
-        response = request.request_head(uri.path)
-        response.code.to_i != STATUS_OK
+    def resize_with_black_background(image, width, height)
+        image.combine_options do |c|
+            c.extent("#{width}x#{height}")
+            c.background("black")
+            c.gravity("center")
+        end
     end
 
     def resize_image(path, output_path, width, height)
-        
-        image = MiniMagick::Image.open(path)
-        if width > image.width && height > image.height
-            image.combine_options do |c|
-                c.extent("#{width}x#{height}")
-                c.background("black")
-                c.gravity("center")
-            end
-        elsif width > image.width || height > image.height
-            image = image.resize("#{width}x#{height}")
-            image.combine_options do |c|
-                c.extent("#{width}x#{height}")
-                c.background("black")
-                c.gravity("center")
+        begin
+            image = MiniMagick::Image.open(path)
+            if width > image.width && height > image.height
+                resize_with_black_background(image, width, height)
+            elsif width > image.width || height > image.height
+                image = image.resize("#{width}x#{height}")
+                resize_with_black_background(image, width, height)
+            else
+                image = image.resize("#{width}x#{height}")
             end
-        else
-            image = image.resize("#{width}x#{height}")
+            image.write("#{output_path}")
+            output_path
+        rescue MiniMagick::Invalid => e
+            raise IOError.new("URL not found or not an image")
+        rescue OpenURI::HTTPError => e
+            raise IOError.new("URL not found or not an image")
+        rescue OpenSSL::SSL::SSLError => e
+            raise IOError.new("Failed to open ssl connection")
+        rescue SocketError => e
+            raise IOError.new("Could not open connection to #{path}")
         end
-        image.write("#{output_path}")
     end
 
     def handle_request(url, width, height)
-    
-        if url.nil? || image_doesnt_exist?(url) # todo check that it is actually and image!
-            [BAD_REQUEST, parse_error_to_json(BAD_REQUEST, "url not found or not an image")]
-        elsif width.nil? || height.nil? || width.to_i <= 0 || height.to_i <= 0
-            [BAD_REQUEST, parse_error_to_json(BAD_REQUEST, "Invalid params")]
+        if url.nil? || width.nil? || height.nil?
+            raise ArgumentError.new("Arguments cannot be nil, got: url=#{url}, width=#{width}, height=#{height}")
+        elsif width.to_i <= 0 || height.to_i <= 0
+            raise ArgumentError.new("Width and height must be positive integers, got: width=#{width}, height=#{height}")
         else
             resize_image(url, @local_path, width.to_i, height.to_i)
```

Comment:
`@local_path` does not solve our concurrency problem.
Think about race condition - 10 concurrent requests arrive at the same time. `@local_path` will get overwritten simultaneously...